### PR TITLE
[TBDGen] Capture segment symbols are defined in & emit tbd-v5 format

### PR DESCRIPTION
**Pull request by cyndyishida**
TBD-v5 is the new json format for TBD files. One can view the format spec by running `man tapi-tbd` on any >=Xcode15 install. Because it captures the _segment_ the symbols are defined in, to restore back information lost compared to binary dylibs, it is also now recorded for TBDGen in the compiler. This is a best effort, as there is no way to definitively know where the symbol ends up until after the linker has assembled the binary.

This patch also updates tests to 
1. add flags for required information to emit the tbd (install name)
2. Use llvm tools that capture relevant information out of TBD files where possible (llvm-readtapi & llvm-nm)

Resolves: rdar://117604275

### CODE REVIEW

**Comment by cyndyishida** [comment]
I expect some tests that invoke the Xcode toolchain linker to fail as Xcode14 is too old to read tbdv5. They should be resolved by: https://github.com/apple/swift/issues/69485 

I can temporarily disable them, but I figure its better to get feedback first.

File: test/TBD/dylib-version-truncation.swift [hunk 1]
```diff
@@ -1,27 +1,35 @@
 // REQUIRES: VENDOR=apple
 // RUN: %empty-directory(%t)
 
-// RUN: %target-swift-frontend -typecheck %s -emit-tbd-path - -tbd-current-version 2.0 | %FileCheck %s --check-prefix TWOPOINTZERO
-// RUN: %target-swift-frontend -typecheck %s -emit-tbd-path - -tbd-current-version 2 | %FileCheck %s --check-prefix TWOPOINTZERO
-// RUN: %target-swift-frontend -typecheck %s -emit-tbd-path - -tbd-current-version 20.10 | %FileCheck %s --check-prefix TWENTYPOINTTEN
+// RUN: %target-swift-frontend -typecheck %s -tbd-install_name some_dylib -emit-tbd-path - -tbd-current-version 2.0 | %FileCheck %s --check-prefix TWOPOINTZERO
+// RUN: %target-swift-frontend -typecheck %s -tbd-install_name some_dylib -emit-tbd-path - -tbd-current-version 2 | %FileCheck %s --check-prefix TWOPOINTZERO
+// RUN: %target-swift-frontend -typecheck %s -tbd-install_name some_dylib -emit-tbd-path - -tbd-current-version 20.10 | %FileCheck %s --check-prefix TWENTYPOINTTEN
```
Comment (xymus) [inline test/TBD/dylib-version-truncation.swift:6]: I notice you're adding many `-tbd-install_name` to tests, is it a requirement to set it explicitly now?

**Review by xymus — APPROVED** [review]
LGTM!

File: test/TBD/dylib-version-truncation.swift (hunk 1 above)
Comment (cyndyishida) [reply test/TBD/dylib-version-truncation.swift:6]: Yes! It is an assert check in LLVM now. https://github.com/llvm/llvm-project/blob/main/llvm/lib/TextAPI/TextStubV5.cpp#L370

**Comment by cyndyishida** [comment]
@swift-ci please test

**Review by tshortli — APPROVED** [review]

**Review by nkcsgexi — APPROVED** [review]
Thank you, Cyndy!

**Comment by cyndyishida** [comment]
@swift-ci please test

**Pull request merged**